src/opustrainer/trainer.py

```diff
@@ -24,6 +28,101 @@ def ignore_sigint():
+            curr: int = inlist[i][0]
+            if curr not in blacklisted:
+                if i < len(inlist) - 1:
+                    peek: int = inlist[i+1][0]
+                    if curr==peek:
+                        blacklisted.add(curr)
+                    else:
+                        new_list.append(inlist[i])
+                else:
+                    new_list.append(inlist[i])
+        return new_list
+
+    src_trg_filtered: List[Tuple[int, int]] = filter_tuples(src_trg)
+    trg_src_filtered: List[Tuple[int, int]] = filter_tuples(trg_src)
+
+    # Now, we are looking for the union of the two.
+    # First, reverse src_trg
+    trg_src_filtered_rereversed: List[Tuple[int, int]] = [(c, d) for d,c in trg_src_filtered]
+
+    # Now convert both to sets and take the union
+    src_trg_set: Set[Tuple[int, int]] = set(src_trg_filtered)
+    trg_src_filtered_rereversed_set: Set[Tuple[int, int]] = set(trg_src_filtered_rereversed)
+
+    return list(src_trg_set & trg_src_filtered_rereversed_set)
+
+# Unpacks a line, removes the alignments, and applies placeholding. Hardcoded for the moment
+# Also applies detokenization on the source side, because getting word alignments for Chinese is otherwise hard
+
+def tag_line(line: str, *, probability: float=0.0, num_tags: int=6, chinese: str=None) -> str:
+    '''Applies tag to words in a line based on alignment info, and then removes the alignment info from the line.
+       This is used to enable terminology support by tagging random words with their translation.
+       eg "I like cake" would become "I like <tag0> gusta </tag0> cake"
+    '''
+    src, trg, alignment = line.strip().split('\t')
+    source = src.split(' ')
+    target = trg.split(' ')
+
+    # Get replacement candidates
+    candidates: List[Tuple[int, int]] = get_placeholding_candidates(alignment)
+
+    # Get list of possible tags. Hardcode for now
+    tags: List[str] = [str(x) for x in range(num_tags)]
+    # Shuffle the list. It's quite slow, but we only have 6 elements so it should be fine
+    # For more information https://stackoverflow.com/questions/10048069/what-is-the-most-pythonic-way-to-pop-a-random-element-from-a-list
+    random.shuffle(tags)
+
+    # Replace each of them with a THRESHOLD probability unless the two words are exactly the same
+    # This is to avoid having numbers trained with placeholders or any other words that are exactly the same
+    for i in range(len(candidates)):
+        if source[candidates[i][0]] != target[candidates[i][1]] and  random.random() < probability:
+            # Try to get a tag if available and tag our data
+            if len(tags) > 0:
+                tag_id = tags.pop()
+                source[candidates[i][0]] = source[candidates[i][0]] + " <tag" + tag_id + "> " + target[candidates[i][1]] + " </tag" + tag_id + ">"
+            else:
+                # We run out of tags so no point of trying to tag anything else
+                break
```

Comment:
It can deal with multi token input because sometimes there's multiple SPM pieces inside a single tag. That being said, it's not explicit, but i verified it with a few examples that it did work in practise. I need to have a better test set to verify it.

Is the shuffle random enough you reckon?